src/XMLHelper.c

```diff
@@ -0,0 +1,383 @@
+            fprintf(stream, "\n");
+        }
+
+        text = RTIXMLUTILSObject_getText(node);
+        if (strlen(text) > 0) {
+            for (i = 0; i < indent + 1 && has_children; i++) {
+                fprintf(stream, "%s", INDENTATION_UNIT);
+            }
+            fprintf(stream, "%s", text);
+            if (has_children) {
+                fprintf(stream, "\n");
+            }
+        }
+
+        XMLHelper_print_node(
+                RTIXMLUTILSObject_getFirstChild(node), 
+                indent + 1, 
+                stream, 
+                0);
+
+        for (i = 0; i < indent && has_children; i++) {
+            fprintf(stream, "%s", INDENTATION_UNIT);
+        }
+        fprintf(stream, "</%s>\n", RTIXMLUTILSObject_getTagName(node));
+
+        if (!is_root) {
+            node = RTIXMLUTILSObject_getNextSibling(node);
+        } else {
+            break;
+        }
+    }
+}
+
+void XMLHelper_pretty_print(FILE *stream, char *string, char *query) 
+{
+    struct RTIXMLUTILSObject *dom_root = NULL, *query_root = NULL;
+
+    RTIXMLUTILSParser_parseString(&dom_root, string);
+    if (query != NULL) {
+        query_root = RTIXMLUTILSObject_getFirstChildWithTag(dom_root, query);
+        if (query_root == NULL) {
+            printf("The queried subtag '%s' couldn't be found! \n", query);
+        } else {
+            XMLHelper_print_node(query_root, 0, stream, 1);
+        }
+    } else {
+        XMLHelper_print_node(dom_root, 0, stream, 1);
+    }
+
+    RTIXMLUTILSParser_freeDom(dom_root);
+}
+
+RTI_Retval XMLHelper_insert_enclosing_tag(
+        char *qos_type, 
+        struct RTIXMLSaveContext *context, 
+        int closing)
+{
+    RTI_Retval result = RTI_ERROR;
+    const char *closing_format = "</%s>\n\0";
+    const char *opening_format = "<%s>\n";
+    const int closing_format_length = 5;
+    const int opening_format_length = 3;
+    int format_length = 0;
+    char *format = NULL;
+
+    if (closing) {
+        format_length = closing_format_length;
+        format = (char *) closing_format;
+    } else {
+        format_length = opening_format_length;
+        format = (char *) opening_format;
+    }
+
+    if (context->fout != NULL) {
+        fprintf(context->fout, format, qos_type);
+        context->outputStringLength += (strlen(qos_type) + format_length);
+    } else if (context->sout != NULL) {
+        if (context->ssize >= (context->outputStringLength + strlen(qos_type) + format_length)) {
+            char *current_position = context->sout;
+            current_position += context->outputStringLength;
+            sprintf(current_position, format, qos_type);
+            context->outputStringLength += (strlen(qos_type) + format_length);
+        } else {
+            goto done;
+        }
+    } else {
+        context->outputStringLength += (strlen(qos_type) + format_length);
+    }
+
+    result = RTI_OK;
+done:
+    return result;
+}
+
+RTI_Retval XMLHelper_dump_datawriter_qos(
```

Comment:
we don't need to do it now, but we can write this logic in a single place and leverage a table-driven approach to provide the tag and the get qos operations. It will look much cleaner and reusable.

Reply:
You mean like a function pointer approach?

Reply:
yes